File: LeetCode/Medium/0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        ret = []

        hq = [(0,0,0)]
        visited = set([(0,0)])

        while k and hq:
            _, i, j = heapq.heappop(hq)

            ret.append([nums1[i], nums2[j]])

            if i + 1 < len(nums1) and (i+1, j) not in visited:
                heappush(hq, (nums1[i+1] + nums2[j], i + 1, j))
                visited.add((i+1, j))
            if j + 1 < len(nums2) and (i, j+1) not in visited:
                heappush(hq, (nums1[i] + nums2[j+1], i, j+1))
                visited.add((i, j+1))
            k -= 1

        return ret

File: LeetCode/Medium/0373-find-k-pairs-with-smallest-sums/0373-find-k-pairs-with-smallest-sums.java
class Solution {
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        PriorityQueue<Integer[]> pq = new PriorityQueue<>((o1, o2) -> {
            return Integer.compare(o1[0], o2[0]);
        });

        List<List<Integer>> ret = new ArrayList();

        for (int i = 0; i < k && i < nums1.length; ++i) pq.offer(new Integer[]{nums1[i] + nums2[0], i, 0});
        
        while (k > 0 && !pq.isEmpty()){
            Integer[] target = pq.poll();
            
            int i = target[1];
            int j = target[2];

            ret.add(Arrays.asList(nums1[i], nums2[j]));
            
            if (j + 1 < nums2.length)
                pq.offer(new Integer[]{nums1[i] + nums2[j+1], i, j+1});
            

            k--;
        }

        return ret;

    }
}